ssl/quic/quic_port.c

```diff
@@ -472,22 +495,49 @@ static QUIC_CHANNEL *port_make_channel(QUIC_PORT *port, SSL *tls, int is_server)
 
     args.port       = port;
     args.is_server  = is_server;
-    args.tls        = (tls != NULL ? tls : port_new_handshake_layer(port));
     args.lcidm      = port->lcidm;
     args.srtm       = port->srtm;
-    if (args.tls == NULL)
+
+    /*
+     * Creating a a new channel is made a bit tricky here as there is a
+     * bit of a circular dependency.  Initalizing a channel requires that
+     * the ch->tls and optionally the qlog_title be configured prior to
+     * initalization, but we need the channel at least partially configured
+     * to create the new handshake layer, so we have to do this in a few steps.
+     */
+
+    /*
+     * start by allocation and provisioning as much of the channel as we can
+     */
+    ch = ossl_quic_channel_alloc(&args);
+    if (ch == NULL)
         return NULL;
 
+    /*
+     * Fixup the channel tls connection here before we init the channel
+     */
+    ch->tls = (tls != NULL) ? tls : port_new_handshake_layer(port, ch);
+
 #ifndef OPENSSL_NO_QLOG
-    args.use_qlog   = 1; /* disabled if env not set */
-    args.qlog_title = args.tls->ctx->qlog_title;
+    /*
+     * If we're using qlog, make sure the tls get further configured properly
+     */
+    ch->use_qlog = 1;
+    if (ch->tls->ctx->qlog_title != NULL) {
+        if ((ch->qlog_title = OPENSSL_strdup(ch->tls->ctx->qlog_title)) == NULL) {
+            OPENSSL_free(ch);
+            return NULL;
+        }
+    }
 #endif
 
-    ch = ossl_quic_channel_new(&args);
-    if (ch == NULL) {
-        if (tls == NULL)
-            SSL_free(args.tls);
-
+    /*
+     * And finally init the channel struct
+     */
+    if (!ossl_quic_channel_init(ch)) {
+        if (ch->tls == NULL)
+            SSL_free(ch->tls);
```

Comment:
I think the if at line 538 should actually read as `if (ch->tls != NULL)`

Reply:
also one more thought: the test at line 538 is not necessary at all because the `SSL_free()` reads as follows:
```
1379 void SSL_free(SSL *s)
1380 {
1381     int i;
1382
1383     if (s == NULL)
1384         return;
1385     CRYPTO_DOWN_REF(&s->references, &i);
```